src/main/flight/autopilot.c

```diff
@@ -91,14 +196,172 @@ void altitudeControl(float targetAltitudeCm, float taskIntervalS, float vertical
+                // if entering poshold from a stable hover, we would only exceed this if IMU was disoriented
+                // if entering poshold at speed, it may overshoot this value and falsely fail, if so need something more complex
+            }
+    
+            vectors_t *vectors[] = { &posHold.NS, &posHold.EW };
+            for (int i = 0; i < 2; i++) {
+                vectors_t *latLong = vectors[i];
+    
+                // separate PID controllers for latitude (NorthSouth or ns) and longitude (EastWest or ew)
+    
+                // ** P **
+                float pidP = latLong->distance * positionPidCoeffs.Kp;
+    
+                // ** I **
+                if (!latLong->isStarting){
+                    // only accumulate iTerm after completing the start phase
+                    // perhaps need a timeout on the start phase ?
+                    latLong->integral += latLong->distance * posHold.gpsDataIntervalS;
+                } else {
+                    // while moving sticks, slowly leak iTerm away, approx 2s time constant
+                    latLong->integral *= leak;
+                }
+                float pidI = latLong->integral * positionPidCoeffs.Ki;
+        
+                // ** D ** //
+                // get change in distance in NS and EW directions from gps.c using the `GPS_distances` function
+                // this gives cleaner velocity data than the module supplied GPS Speed and Heading information
+                float velocity = (latLong->distance - latLong->previousDistance) * posHold.gpsDataFreqHz; // cm/s, minimum step 11.1 cm/s
+                latLong->previousDistance = latLong->distance;
+    
+                float acceleration = (velocity - latLong->previousVelocity) * posHold.gpsDataFreqHz;
+                latLong->previousVelocity = velocity;
+        
+                // scale and filter - filter cutoffs vary during the startup phase
+                float pidD = velocity * positionPidCoeffs.Kd;
+                pt1FilterUpdateCutoff(&latLong->velocityLpf, posHold.pt1Gain);
+                pidD = pt1FilterApply(&latLong->velocityLpf, pidD);
+    
+                float pidA = acceleration * positionPidCoeffs.Kd;
+                pt2FilterUpdateCutoff(&latLong->accelerationLpf, posHold.pt1Gain);
+                pidA = pt2FilterApply(&latLong->accelerationLpf, pidA);
+        
+                // limit sum of D and A because otherwise can be too aggressive when starting at speed
+                const float maxDAAngle = 35.0f; // limit in degrees; arbitrary.  20 is a bit too low, allows a lot of overshoot
+                // an angle of more than 35 degrees is achieved as P and I grow
+                // ** todo = should this be half of the user-configurable angle_limit?  Or fixed?
+                const float pidDA = constrainf(pidD + pidA, -maxDAAngle, maxDAAngle);
```

Comment:
It may be better to limit D+A vector - limiting each component separately will change PID (2d) direction

Reply:
Yes this is a concern that I had, but wasn't sure how to most effectively solve that concern.
In practice it probably doesn't matter much when one axis is dominant, but could be significant for hard diagonal stops.
What would you suggest?

Reply:
```
vector2_f pidDA2d;
float mag = vector2Norm(&pidDA2d);
if (mag > maxDAAngle) {
   vector2Scale(&pidDA2d, maxDAAngle / mag);
}